Improve generality of SpriteCache AdjustFrame.

pchote opened the pull request:
This PR makes some tweaks to #21369 to restore support for TDHD, and makes some small cleanups to D2kSpriteSequence.

pchote committed:
Remove obsolete D2k ResolveSprites override.

pchote committed:
Fix D2kSpriteSequence namespace.

RoosterDragon commented on OpenRA.Game/Graphics/SpriteCache.cs:
```diff
@@ -82,7 +87,7 @@ public void LoadReservations(ModData modData)
 				int FrameIndex,
 				bool Premultiplied,
-				Func<ISpriteFrame, ISpriteFrame> AdjustFrame,
+				Func<ISpriteFrame, int, int, ISpriteFrame> AdjustFrame,
```
I think introducing a custom delegate could be nice here, e.g.
`public delegate ISpriteFrame AdjustFrame(ISpriteFrame input, int index, int total);`

That would allow for some documentation as well as avoiding having to change the signature in so many places in the future should we ever tweak it again.

RoosterDragon commented on OpenRA.Game/Graphics/SpriteCache.cs:
```diff
@@ -73,6 +73,11 @@ static ISpriteFrame[] GetFrames(IReadOnlyFileSystem fileSystem, string filename,
 
+		public ISpriteFrame[] LoadFramesUncached(string filename)
+		{
+			return GetFrames(fileSystem, filename, loaders, out _);
```
Maybe `GetFrames` should remove the `out TypeDictionary metadata` parameter since neither caller uses it?

RoosterDragon reviewed (commented):
Minor suggestions.

pchote commented:
Updated

pchote committed:
Call AdjustFrame with frame index and total.

pchote committed:
Add SpriteCache.LoadFramesUncached.

This allows users (currently TDHD) to load ISpriteFrames
directly, without them being stored in the cache.

PunkPun commented:
needs a rabase

pchote commented:
I'd prefer to cncbase, but done 😉

PunkPun commented on OpenRA.Game/Graphics/SpriteCache.cs:
```diff
@@ -49,30 +50,34 @@ public sealed class SpriteCache : IDisposable
 		}
 
-		static ISpriteFrame[] GetFrames(IReadOnlyFileSystem fileSystem, string filename, ISpriteLoader[] loaders, out TypeDictionary metadata)
+		static ISpriteFrame[] GetFrames(IReadOnlyFileSystem fileSystem, string filename, ISpriteLoader[] loaders)
```
while this probably isn't the right place to get metadata, it seems to be the only place. I recall @IceReaper having issues about metadata being disposed

IceReaper replied:
Yeah, kknd has per frame metadata like 2d bone positions. Thats required to make the actors look like they should look when animating them

PunkPun commented:
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/5c9bd1ee262028e247d34075d2f7756535d5f560...d300295c99d0941aca826c9bc58dc4cade1c48a4)
